### ISSUE TITLE
Documentation 

### ISSUE THREAD

**Issue by argentum2f** [nchopin/particles#8]
Firstly, thanks for contributing this, your efforts are appreciated!

I've been trying to test this library and struggled to make headway due to 
1) Missing documentation (not self-contained)

2) Example code is often non-functional (missing imports/definitions that are critical to understanding it), missing arguments in the docstrings and completely missing docstrings (*every* function and class should have one).

3) The library seems designed to process complete data sets, and can't be used as is for live filtering?

Some examples of the above:
**1.**  The documentation often refers to 'the book', and isn't very well self contained. The documentation also often explains things using a particular math notation, without defining that notation. In general it's always better when possible with technical writing to use full words in addition to symbols, rather than replacing words with symbols (e.g. rather than saying "M0(dx0) is the N(0,1) distribution", you could say "The initial distribution M0(dx0) is in this case the normal distribution N(0,1)". That way if it isn't immediately apparent what 'M' or 'N' are meant to be to someone who just picked up the library or doesn't have the book handy, there is enough info to go off of, or to google more information (e.g. wikipedia `normal distribution`, etc. if needed).

Examples of missing docstrings: 
All of these functions from `particles.core.SMC` are missing docstrings: (compute_summaries(),
generate_particles(), next(), resample_move(), resample_move_qmc()). Some of them, like next(), may be self-explanatory for an experienced python user, but I have no idea what  resample_move_qmc() does.

Missing explanation of logG:
Neither the SMC or FeynmanKac classes say they require the logG function to be implemented - just M0, M and G (see here: 
https://particles-sequential-monte-carlo-in-python.readthedocs.io/en/latest/_autosummary/particles.core.FeynmanKac.html#particles.core.FeynmanKac )
but then I get 
`NotImplementedError: method/property logG missing in class my_fk_model`
logG would, presumably, be log(G), which the documentation does state - but nothing more. If it's just log(G) why is it needed at all (i.e. if G is given, can't you just compute log(G)?)? In the few examples I could find where logG is used, it is implemented as a piece wise function that's either 0 or -infinity and I'm not sure why. The documentation should explain what the purpose of logG (in addition to G) is, and that it is *required* along with the other functions mentioned.

**2.** If you look at the intro in README.md, the example code does not include import statements -what is `ssm`, what is `dists` etc. I had to find elsewhere in the docs at some point you did some `import ... as ssm` so I could translate what you meant. 

**3.** I've been trying to work around this by adding an `observations` list to my FeynmanKac class, so I can add an observation and update T before calling next(), then use the time step parameter `t` to retrieve the latest observation for weighting functions. It feels kinda hackish and like I'm not using the library as intended.

I'm coming from having just tested `pfilter`. The documentation and approach of that one made a lot more sense to me, but it was pretty lacking in features, so I was hoping `particles` would work out a bit better. I'm not a filtering expert and have only a basic knowledge of particle filters. I'm not trying to become an expert or anything. Just wanted to get something working ,then test out different types of algorithms/resamplings, etc

**Comment by nchopin** [comment]
Hi, thanks for your feedback. 
First may I ask whether you had a look at the notebook tutorials, as recommended in the front page of the documentation? It should be hopefully give a better idea of how the package is meant to be used. If not, please let me know. 

Regarding your specific points: 

1. 

 - the book: well, this package was developed to supplement a book... which will be available shortly :-) The idea is to refer to the book for more advanced notions and explanations of the math principles behind each method. Same things for the notations, which are also (partly?) explained in the notebooks. (I need to double-check.) 

 - missing docstrings: please note that most of these methods you mention are *private* (not supposed to be used directly by the user). I could use the _method convention to make this more explicit, maybe. Method __next__ *is* (briefly) documented, and, more importantly, the fact that SMC objects are iterators is explained in the class docstring. (Method next is for Python 2 compatibility only, and is the same thing as __next__.) 

 - G vs logG: good catch, there is **no** function G, you are expected to supply a function logG which returns the log of G_t. Now fixed (experimental branch). All calculations in particles are done on the log scale. 

2. OK, now fixed (experimental branch). 

3. sorry, to work around what exactly? I guess you are referring to your point 3 (processing complete datasets)? Well, yes, initially, the SMC class was designed to process a complete dataset (through method run()), but now that you can iterate manually a SMC object, nothing prevents you to add data on the go. The way you do for instance seems fine. Maybe something more elegant would be to accept the user to provide an iterable as the *data*? I need to think about it. 

I will have a look at pfiler at some point. If you find other specific parts of the documentation, please let me know.

**Comment by argentum2f** [comment]
I don't think `next()` really needs any explanation (sorry that came across wrong), but yes, I was confused about the internal functions. Adding a prefix `_` I think would be helpful, or just stating 'internal' in the docstring or some such, so it's clear it's something that can be ignored by the user.

I did look through the notebook tutorials, but they (and the docs in general) depend fairly heavily on field specific jargon and notation, and I had a difficult time trying to figure out even what class was even applicable to my problem of interest. My intro to particle filters (quite recently) was this video:
https://www.youtube.com/watch?v=aUkBa1zMKv4 From that it's obvious to me what a particle filter is (or at least a particular kind) and how it could work. I have been able to successfully implement one that works well for my problem of interest. I'm now interested in learning more and improving on my approach which is what brought me to your library, which looks like a great way to learn more and test out more algorithms.

The problem is, I don't know what 'Feynman-Kac' is, or 'SMC', 'QMC', 'HMMM' etc. are so I've had a hard time finding a good entry point. The docs expand the acronyms sometimes, and give an equation but generally lack the two things I'm looking for: an easy to understand concise conceptual explanation of a model including what it does and what problem it is meant to solve. The docs definitely tend more towards the academic/rigorous/pedantic than toward the applied/easily digestible (which is ok, if that's what you're going for).

Presumably, the explanations I'm looking for are in the book. Is it coming out anytime soon, and can you recommend an alternative in the meantime?

*TLDR;* I struggled to get a filter actually running using just the code/docs (though I have been able to implement similar filters from other simpler packages), but that's ok if I'm not really the intended audience. I'd probably be better off going through the book first before trying to dive straight in and apply your lib to a real problem.

**Comment by nchopin** [comment]
Feynman-Kac: you can skip that, if you are only interested in particle filtering for a given model; simply use the state_space_models.Bootstrap class as shown in the tutorials. 

Book: should be available in a few months (fingers crossed). 

Ok, I keep the issue open, to remind me to review all the documentation, and update it to make it more friendly to people less familiar with the terminology. Perhaps I will add more details in the tutorials for instance. Thanks again for taking the time to give your perspective, it's always helpful to see how a given package/book/paper is received from different ppl.

**Comment by argentum2f** [comment]
Ok, sounds good. Thanks again for making this publicly available!

**Comment by lionfish0** [comment]
I think more examples would be useful, of simple things people want to do:

1) Examples with a state space with more than one dimension.
2) Examples where you want to make predictions where there isn't data?

**Comment by nchopin** [comment]
I will close this issue since: 
1. the book is now available
2. I added a tutorial on multivariate (or more generally, non-trivial) state-space models, see [here](https://particles-sequential-monte-carlo-in-python.readthedocs.io/en/latest/notebooks/complicated_state_space_models.html)
But feel free to open a new issue if you feel some other point is not properly documented.
